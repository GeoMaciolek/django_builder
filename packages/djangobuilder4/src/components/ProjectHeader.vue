<script setup lang="ts">
import {
  DjangoApp,
  DjangoVersion,
  type DjangoProject,
} from "@djangobuilder/core";
import { useUserStore } from "../stores/user";
import { storeToRefs } from "pinia";
import { updateProject } from "@/api";
import EditableText from "@/widgets/EditableText.vue";
import EditableBoolean from "@/widgets/EditableBoolean.vue";
import EditableChoice from "@/widgets/EditableChoice.vue";
import { deleteApp } from "@/firebase";
import ConfirmableButton from "@/widgets/ConfirmableButton.vue";

const props = defineProps<{
  project: DjangoProject;
}>();

const DjangoVersionChoices = {
  [`${DjangoVersion.DJANGO2}`]: DjangoVersion.DJANGO2.toString(),
  [`${DjangoVersion.DJANGO3}`]: DjangoVersion.DJANGO3.toString(),
  [`${DjangoVersion.DJANGO4}`]: DjangoVersion.DJANGO4.toString(),
};

const userStore = useUserStore();
const { getProjectId, getAppId } = storeToRefs(userStore);
const projectid = getProjectId.value(props.project);

function handleDeleteApp(app: DjangoApp) {
  const projectid = getProjectId.value(props.project);
  const appid = getAppId.value(app);
  if (projectid && appid) {
    deleteApp(projectid, appid);
  }
}
</script>

<template>
  <div class="project-header-wrapper">
    <router-link :to="{ name: 'project', params: { id: projectid } }">{{
      project.name
    }}</router-link>

    <table cellpadding="0" cellspacing="0">
      <tr>
        <td>Description</td>
        <td>
          <EditableText
            :value="props.project.description"
            v-on:update="
              (description: string) => updateProject(props.project, { description })
            "
          >
            {{ props.project.description }}
          </EditableText>
        </td>
      </tr>
      <tr>
        <td>django version</td>
        <td>
          <EditableChoice
            :value="String(project.version)"
            :choices="DjangoVersionChoices"
            display_as="radio"
            v-on:update="(django_version: string) => updateProject(props.project, { django_version })"
          />
        </td>
      </tr>
      <tr>
        <td>htmx</td>
        <td>
          <EditableBoolean
            :value="props.project.htmx"
            v-on:update="(htmx) => updateProject(props.project, { htmx })"
          />
        </td>
      </tr>
      <tr>
        <td>channels</td>
        <td>
          <EditableBoolean
            :value="props.project.channels"
            v-on:update="
              (channels) => updateProject(props.project, { channels })
            "
          />
        </td>
      </tr>
      <tr>
        <td>Apps</td>
        <td>
          <div v-for="app in project.apps" :key="app.name">
            {{ app.name }}
            <ConfirmableButton
              :message="`Are you sure you wish to delete the app '${app.name}'?`"
              @confirm="handleDeleteApp(app)"
            />
          </div>
        </td>
      </tr>
    </table>
  </div>
</template>

<style scoped>
a {
  display: inline-block;
  margin: 10px 0px;
}
.description {
  overflow-y: hidden;
  margin-bottom: 4px;
}
.project-header-wrapper {
  border: 1px dotted lightgray;
  margin: 4px;
  padding: 10px;
  display: flex;
  flex-direction: column;
}
table {
  table-layout: fixed;
}
td {
  width: 50%;
  vertical-align: top;
}
.app {
  padding-right: 5px;
}
.delete-app-button {
  margin: 0;
  padding: 0 4px;
  color: red;
}
</style>
