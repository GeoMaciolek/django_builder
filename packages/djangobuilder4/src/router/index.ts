import { createRouter, createWebHistory } from "vue-router";
import LoginView from "../views/LoginView.vue";

import { auth } from "../firebase";
import type { User } from "@firebase/auth";

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: "/login",
      name: "login",
      component: LoginView,
    },
    {
      path: "/",
      name: "projects",
      component: () => import("../views/ProjectsView.vue"),
      meta: {
        requiresAuth: true,
      },
    },
    {
      path: "/project/:id",
      name: "project",
      component: () => import("../views/ProjectView.vue"),
      props: true,
      meta: {
        requiresAuth: true,
      },
    },
    {
      path: "/reset_password",
      name: "reset_password",
      component: () => import("../views/ResetPassword.vue"),
    },
    {
      path: "/action",
      name: "action",
      component: () => import("../views/ActionView.vue"),
    },
    {
      path: "/reset_password_action",
      name: "reset_password_action",
      component: () => import("../views/ResetPasswordAction.vue"),
      props: true,
    },
    {
      path: "/verify_email_action",
      name: "verify_email_action",
      component: () => import("../views/VerifyEmailAction.vue"),
      props: true,
    },
    {
      path: "/unverified",
      name: "unverified",
      component: () => import("../views/UnverifiedUser.vue"),
      props: true,
      meta: {
        requiresAuth: true,
      },
    },
  ],
});

function userVerified(user: User) {
  const githubVerified = user.providerData.find(
    (provider) => provider.providerId === "github.com"
  );
  return githubVerified !== undefined || user.emailVerified || user.isAnonymous;
}

router.beforeEach((to, from, next) => {
  //console.debug("from", from);
  //console.debug("to", to);
  const currentUser = auth.currentUser;
  const requiresAuth = to.matched.some((record) => record.meta.requiresAuth);

  if (requiresAuth && !currentUser) {
    next({ name: "login" });
  } else if (currentUser) {
    const verified = userVerified(currentUser);
    // console.log('Current User verified', verified, 'requesting', to.name)
    if (verified && to.name == "login") {
      // console.log('User is verified and trying to access the login page so send to home page')
      next({ name: "projects" });
    } else if (verified && to.name == "unverified") {
      // console.log('User is verified and trying to access the unverified page so send to home page')
      next({ name: "projects" });
    } else if (
      !verified &&
      to.name != "unverified" &&
      to.name != "verify_email_action"
    ) {
      // console.log('2. User is unverified and trying to access another page so send to unverified page')
      next({ name: "unverified" });
    } else {
      // console.log('User sent to page', to.name)
      next();
    }
  } else {
    // console.log('Anonymous User sent to page that require no auth', to.name)
    next();
  }
});

export default router;
