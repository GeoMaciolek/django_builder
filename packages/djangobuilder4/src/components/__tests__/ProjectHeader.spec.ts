import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import ProjectHeader from "../ProjectHeader.vue";
import type { Project } from "@/types";
import router from "../../router";

describe("ProjectHeader", () => {
  it("renders properly", () => {
    const project = {
      id: "123456",
      name: "Project 1",
      description: "",
      owner: "123",
      channels: true,
      apps: {},
      django_version: 3,
    } as Project;
    const wrapper = mount(ProjectHeader, {
      global: {
        plugins: [router],
      },
      props: { project: project },
    });
    expect(wrapper.text()).toMatchSnapshot();
    expect(wrapper.text()).toContain("Project 1");
  });
});
