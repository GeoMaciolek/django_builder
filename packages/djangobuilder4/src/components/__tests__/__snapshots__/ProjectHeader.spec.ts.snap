// Vitest Snapshot v1

exports[`ProjectHeader > renders properly 1`] = `"Project 1"`;
